packages/core/core/src/Validation.js

```diff
@@ -0,0 +1,99 @@
+// @flow strict-local
+import nullthrows from 'nullthrows';
+import type {AssetRequest, ParcelOptions} from '@parcel/types';
+
+import path from 'path';
+import Cache from '@parcel/cache';
+import {resolveConfig} from '@parcel/utils';
+
+import type Config from './public/Config';
+import {report} from './ReporterRunner';
+import InternalAsset from './Asset';
+import type {NodeId, ConfigRequest} from './types';
+import {MutableAsset} from './public/Asset';
+import summarizeRequest from './summarizeRequest';
+
+export type ValidationOpts = {|
+  request: AssetRequest,
+  loadConfig: (ConfigRequest, NodeId) => Promise<Config>,
+  parentNodeId: NodeId,
+  options: ParcelOptions
+|};
+
+export default class Validation {
+  request: AssetRequest;
+  configRequests: Array<ConfigRequest>;
+  loadConfig: ConfigRequest => Promise<Config>;
+  options: ParcelOptions;
+  cache: Cache;
+  impactfulOptions: $Shape<ParcelOptions>;
+
+  constructor({request, loadConfig, parentNodeId, options}: ValidationOpts) {
+    this.request = request;
+    this.configRequests = [];
+    this.loadConfig = configRequest => {
+      this.configRequests.push(configRequest);
+      return loadConfig(configRequest, parentNodeId);
+    };
+    this.options = options;
+  }
+
+  async run(): Promise<void> {
+    if (this.request.filePath.includes('node_modules')) return;
+
+    report({
+      type: 'validation',
+      request: this.request
+    });
+
+    this.cache = new Cache(this.options.outputFS, this.options.cacheDir);
+
+    let asset = await this.loadAsset();
+    let configRequest = {
+      filePath: this.request.filePath,
+      meta: {
+        actionType: 'validation'
+      }
+    };
+
+    let config = await this.loadConfig(configRequest);
+    let parcelConfig = nullthrows(config.result);
+
+    let validators = await parcelConfig.getValidators(this.request.filePath);
+    for (let validator of validators) {
+      await validator.validate({
+        asset: new MutableAsset(asset),
```

Comment:
probably shouldn't be mutable for validators

Reply:
Ow right, this was just copy-paste will change it to Asset